integrate kronos-android ntp clock library into project

rasprague opened the pull request:
integrate ntp clock library into app, make "Share a Positive Diagnosis" date picker use ntp time

rasprague committed:
add Kronos ntp clock lib

add Kronos ntp clock lib

change DiagnosisKeysToken to use ntp time

rasprague committed:
make app context injectable, make NTPTime use it

make app context injectable, make NTPTime use it

removed app context access hack from BaseCovidWatchApplicaiton

rasprague commented:
asana task https://app.asana.com/0/1188301658055194/1188820238689794/f

Apisov commented on app/src/main/java/org/covidwatch/android/data/NTPTime.kt:
```diff
@@ -0,0 +1,20 @@
+import org.koin.core.KoinComponent
+import org.koin.core.inject
+
+class NTPTime {
```
A separate wrapper is nice but having a static reference to a context is a bad thing that leads to memory leaks.

In `Koin` there is `androidApplication()` that returns `applicationContext`, have a look at any wrapper that use `androidApplication()` in `appModule`. For example `Notifications` -- we need something like that and then we will be injecting `NtpTime` anywhere we need it.

Apisov replied:
Also, to make it clear -- I suggest to rename this class to `NtpTime` because of the [guidelines ](https://developer.android.com/kotlin/style-guide#camel_case)

rasprague replied:
removed static reference to context, changed so it's passed into the constructor and never stored

renamed file

rasprague replied:
Made NtpClock injectable. Injected into BaseCovidWatchApplication and started properly from onCreate()

rasprague replied:
modified VerifyPositiveDiagnosisFragment to use NtpTime for date pickers

Apisov commented on app/src/main/java/org/covidwatch/android/data/diagnosiskeystoken/DiagnosisKeysToken.kt:
```diff
@@ -4,13 +4,14 @@ import androidx.room.Entity
     @PrimaryKey val token: String,
     val exposureConfiguration: CovidExposureConfiguration,
-    val providedTime: Long = System.currentTimeMillis(),
+    val providedTime: Long = NTPTime.currentTimeMillis(),
```
We would to remove default value for this property and use injected `NtpTime` to use a proper time. But having here a proper world time is not critical, it's not even used anywhere for now. 

So we could leave the `System` time here because this class might be deleted in the near future when we update EN version.

rasprague replied:
change reverted back to System call

Apisov commented on app/src/main/java/org/covidwatch/android/di/AppContext.kt:
```diff
@@ -0,0 +1,8 @@
+import android.content.ContextWrapper
+
+// exists solely to make the Android applicationContext injectable
+class AppContext(context : Context) : ContextWrapper(context) {
```
Remove it please since we won't use the implementation of `NTPTime` with `companion object`.

rasprague replied:
file removed

Apisov reviewed (changes requested):
It's nice start but according to README from https://github.com/lyft/Kronos-Android
we need to init this lib in the `onCreate` of the `Application` class. So when the wrappers is done we would need to inject it in the `Application` class to call `syncInBackground` method.

Also, to fix the [problem](https://app.asana.com/0/1188301658055194/1188820238689794) we need to limit the date pickers on the verification screen to real dates. 

So we would to rework it here:
https://github.com/covidwatchorg/covidwatch-android-en/blob/21cb2a7ff00504e84382a9d3fd05ceafcd8bc63b/app/src/main/java/org/covidwatch/android/ui/reporting/VerifyPositiveDiagnosisFragment.kt#L122-L128

rasprague committed:
refactor so NtpTime is injectable, change VerifyPositiveDiagnosisFragment to use NtpTime for date pickers

refactor so NtpTime is injectable, change VerifyPositiveDiagnosisFragment to use NtpTime for date pickers

remove AppContext class

properly start ntpTime in Application onCreate()

rasprague committed:
remove unused function from NtpTime

remove unused function from NtpTime

rasprague committed:
remove unused import

remove unused import

rasprague committed:
remove minor change to DiagnosisKeysToken.kt to remove from this commit

removed trailing newline (the only change to this file) to remove from this commit

rasprague commented on app/src/main/java/org/covidwatch/android/data/diagnosiskeystoken/DiagnosisKeysToken.kt:
```diff
@@ -13,4 +13,4 @@ data class DiagnosisKeysToken(
     val potentialExposure: Boolean = false
-)
\ No newline at end of file
+)
```
looks like the github editor enforces the final newline haha

rasprague committed:
remove DiagnonsisKeysToken.kt from commit

remove DiagnonsisKeysToken.kt from commit

Apisov commented on app/build.gradle:
```diff
@@ -288,6 +288,10 @@ dependencies {
     testImplementation "org.koin:koin-test:$koin_version"
+
+    // Kronos ntp clock
+    def kronos_latest_version = "0.0.1-alpha09"
```
Last dependencies are preserved for the test flavored so let's move this before the test dependencies.

rasprague replied:
dependency line moved

Apisov commented on app/src/main/java/org/covidwatch/android/data/NtpTime.kt:
```diff
@@ -0,0 +1,19 @@
+    fun syncInBackground() = kronosClock.syncInBackground()
+
+    fun nowAsLocalDate() : LocalDate =
+        Instant.ofEpochMilli(currentTimeMillis()).atZone(ZoneId.systemDefault()).toLocalDate()
```
`ZoneId.systemDefault()` introduces the same problem that we tried to eliminate -- it tights the time of the app to the users' settings. I was able to reproduce a one day shift with a timezone that have the next day relatively to mine. It's not that big as before but still a problem.

One way to resolve is just to remove `atZone(ZoneId.systemDefault())`.

rasprague replied:
removed notion of LocalDate from NtpTime to avoid confusion / simplify the code. Sticking to returning Instants instead

Apisov commented on app/src/main/java/org/covidwatch/android/ui/reporting/VerifyPositiveDiagnosisFragment.kt:
```diff
@@ -119,13 +121,13 @@ class VerifyPositiveDiagnosisFragment :
         val constraints = CalendarConstraints.Builder()
 
-        val twoWeeksAgo = LocalDate.now().plusDays(-14).toInstant()
+        val twoWeeksAgo = ntpTime.nowAsLocalDate().plusDays(-14).toInstant()
```
There is a problem with this solution.
Even if the dates that can be selected are correct the date picker shows the current month from the device's settings. 

I tried to use selection to current day when we open the dialog but it doesn't help.
@rasprague  Could you try to investigate if that somehow possible to fix with `MaterialDatePicker`?

rasprague replied:
adding a setOpenAt() to the date picker constraints fixes this

Apisov reviewed (changes requested):
Thanks for changes now it's better but still some issues need to be resolved.

I double checked the library and it's quite old but the version of it still still super alpha which suggest Lyft doesn't really support it. There are a few issues in their repo but we don't really need a high accuracy  from the lib -- we need good enough to sync the days not minutes. 
Another old lib that I remember is https://github.com/instacart/truetime-android
but it looks even more abandoned so no big difference I guess.

Also, I realized that introducing NTP time, in the ideal case, would need more changes than here...
We probably need to change most of `Instant.now()` or any `now()` requests but this is too big changes that would need to replace all `java.Time` API. We need to think on this before changing everything.

rasprague committed:
removed "LocalTime" from NtpTime, add plus/minus days extenstion to Instant

removed "LocalTime" from NtpTime

add plus/minus days extenstion to Instant

rasprague committed:
force date picker to open on today's month

force date picker to open on today's month

Apisov commented on app/build.gradle:
```diff
@@ -290,4 +294,4 @@ dependencies {
 
-apply plugin: 'com.google.gms.google-services'
\ No newline at end of file
+apply plugin: 'com.google.gms.google-services'
```
Let's use the same formatter from AS in the future to avoid this kind of changes.

rasprague replied:
agreed

Apisov commented on app/src/main/java/org/covidwatch/android/extension/Instant.kt:
```diff
@@ -0,0 +1,9 @@
+
+import java.time.Instant
+
+// 1d * (24h/d) * (60m/h) * (60s/m)
```
Creating extensions is nice but why then don't use existing API inside of them?
Like this:
```Kotlin
minus(15, ChronoUnit.DAYS)
```
```Kotlin
add(Duration.ofDays(15))
```

rasprague replied:
changed

Apisov commented on app/src/main/java/org/covidwatch/android/ui/reporting/VerifyPositiveDiagnosisFragment.kt:
```diff
@@ -119,15 +121,16 @@ class VerifyPositiveDiagnosisFragment :
-        constraints.setValidator(BaseDateValidator { it in fromWhen..untilNow })
+        constraints
+            .setValidator(BaseDateValidator { it in fromWhen..untilNow })
+            .setOpenAt(untilNow)
```
Nice 👍

Apisov commented on app/src/main/java/org/covidwatch/android/ui/reporting/VerifyPositiveDiagnosisFragment.kt:
```diff
@@ -119,15 +121,16 @@ class VerifyPositiveDiagnosisFragment :
         val fromWhen = dayInPast ?: twoWeeksAgo
 
-        val untilNow = LocalDate.now().toInstant().toEpochMilli()
+        val untilNow = ntpTime.nowAsInstant().toEpochMilli()
```
It looks like the conversion is redundant because we use the milliseconds anyway. We could make `NtpTime.currentTimeMillis` public and use it directly here.

rasprague replied:
calling currentTimeMillis() directly

Apisov reviewed (changes requested):
Almost done, just a few suggestions. But the concern about this PR is still valid -- should we use this NTP adjusted time across the whole app or only in the date pickers.

rasprague committed:
minor changes / cleanup

minor changes / cleanup
